File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    let max = 0;
    
    const func = (arr, idx, cnt, hp) => {
        const fixed = arr[idx];
        const rest = [...arr.slice(0,idx), ...arr.slice(idx+1)];
        hp -= fixed[1];
        
        // 모든 던전 돌았으면 리턴
        if (arr.length===1) {
            max = Math.max(max, cnt);
            return;
        }
        
        // 다음 던전 선택
        for (let i=0; i<rest.length; i++) {
            if (hp>=rest[i][0]) func(rest, i, cnt+1, hp);
            else max = Math.max(max, cnt);
        }
    };
    
    for (let i=0; i<dungeons.length; i++) {
        func(dungeons, i, 1, k);
    };
    
    return max;
}

File: 프로그래머스/2/87946. 피로도/피로도.py
from collections import deque
import copy

def solution(k, dungeons):
    max_cnt = 0
    
    # BFS로 탐색 후 max_cnt 업데이트
    visited = [0]*len(dungeons)
    queue = deque()
    queue.append((0,k,visited))
    while queue:
        cnt,hp,visit = queue.popleft()
        
        if 0 not in visit:
            max_cnt = max(max_cnt, cnt)
        
        flag = False
        for i in range(len(visit)):
            if visit[i]==0 and hp>=dungeons[i][0]:
                if flag==False:
                    flag = True
                arr = copy.deepcopy(visit)
                arr[i]=1
                queue.append((cnt+1, hp-dungeons[i][1],arr))
        if flag==False:
            max_cnt = max(cnt, max_cnt)
    return max_cnt